README.md
# atm-hardware

ATM Hardware, used by [Electron ATM](https://github.com/timgabets/electron-atm) application.

## To use:
TODO
```javascript
const ATMHardwareService = require('atm-hardware');
var hw = new ATMHardwareService();

```

CHANGELOG
October 2017
	0.2.3 	initiailzing cassettes with Good state supply
	0.2.2 	initializing cassette counters during init()
	0.2.1 	initCassetteCounters()
	0.2.0	fixed npm module name ('atm-hardware')
	0.1.9 	getHardwareFitness()
	0.1.8 	setDeviceFitness()
	0.1.7 	setDeviceFitness()
	0.1.6	setDeviceSupply()
	0.1.5 	getSuppliesStatus()
	0.1.4 	init()
	0.1.3 	getReleaaseNumber()
	0.1.2 	getHardwareID()
	0.1.1 	Added dummy test cases
	0.1.0 	Initial commit

spec/ATMHardwareSpec.js
describe('ATMHardwareService', function(){
  const ATMHardwareService = require('../atmhardware.js');
  var hw;

  beforeEach(function() {
    hw = new ATMHardwareService();
  });

  describe('initCassetteCounters()', function(){
    it('should init cassette counters with default values', function(){
      hw.devices = {
        'Cassette 1': {},
        'Cassette 2': {},
        'Cassette 3': {},
        'Cassette 4': {},
      };

      [ 'Cassette 1', 'Cassette 2', 'Cassette 3', 'Cassette 4'].forEach(cassette => {
        expect(hw.devices[cassette].counters).toBeUndefined();
      });

      hw.initCassetteCounters();

      [ 'Cassette 1', 'Cassette 2', 'Cassette 3', 'Cassette 4'].forEach(cassette => {
        expect(hw.devices[cassette].counters).toEqual({ 'loaded': 0, 'dispensed': 0, 'rejected': 0 }); 
      });
    });

    it('should init cassette counters with non-default values', function(){
      hw.devices = {
        'Cassette 1': {},
        'Cassette 2': {},
        'Cassette 3': {},
        'Cassette 4': {},
      };

      [ 'Cassette 1', 'Cassette 2', 'Cassette 3', 'Cassette 4'].forEach(cassette => {
        expect(hw.devices[cassette].counters).toBeUndefined();
      });

      hw.initCassetteCounters(5000, 7, 2);

      [ 'Cassette 1', 'Cassette 2', 'Cassette 3', 'Cassette 4'].forEach(cassette => {
        expect(hw.devices[cassette].counters).toEqual({ 'loaded': 5000, 'dispensed': 7, 'rejected': 2 }); 
      });
    });
  });

  describe('init()', function(){
    it('should get init devices', function(){
      expect(hw.devices).toEqual({});
      hw.init();

      var devices = {
        'High Order Communication': { supply: 'Not configured', fitness: 'No error' },
        'System Disk': { supply: 'Not configured', fitness: 'No error' },
        'Magnetic Card Reader': { supply: 'Not configured', fitness: 'No error' },
        'Cash Handler': { supply: 'Not configured', fitness: 'No error' },
        'Depository': { supply: 'Not configured', fitness: 'No error' },
        'Encryptor': { supply: 'Not configured', fitness: 'No error' },
        'Security Camera': { supply: 'Not configured', fitness: 'No error' },
        'Door Access': { supply: 'Not configured', fitness: 'No error' },
        'Flex Disk': { supply: 'Not configured', fitness: 'No error' },
        'Statement Printer': { supply: 'Not configured', fitness: 'No error' },
        'Signage Display': { supply: 'Not configured', fitness: 'No error' },
        'System Display': { supply: 'Not configured', fitness: 'No error' },
        'Media Entry Indicators': { supply: 'Not configured', fitness: 'No error' },
        'Coin Dispensing Module Tamper Indication': { supply: 'Not configured', fitness: 'No error' },
        'Voice Guidance System': { supply: 'Not configured', fitness: 'No error' },
        'Bunch Note Acceptor': { supply: 'Not configured', fitness: 'No error' },
        'Cheque Processor': { supply: 'Not configured', fitness: 'No error' },
        'Card Capture Bin': { supply: 'Not configured', fitness: 'No error' },
        'Cash Handler Reject Bin': { supply: 'Not configured', fitness: 'No error' },
        'Deposit Bin': { supply: 'Not configured', fitness: 'No error' },
        'Receipt Printer': { supply: 'Not configured', fitness: 'No error' },
        'Journal Printer': { supply: 'Not configured', fitness: 'No error' },
        'Night Safe': { supply: 'Not configured', fitness: 'No error' },
        'Cassette 1': { 
          supply: 'Good state', 
          fitness: 'No error',
          counters: { loaded: 0, dispensed: 0, rejected: 0 }
        },
        'Cassette 2': { 
          supply: 'Good state', 
          fitness: 'No error',
          counters: { loaded: 0, dispensed: 0, rejected: 0 }
        },
        'Cassette 3': { 
          supply: 'Good state', 
          fitness: 'No error',
          counters: { loaded: 0, dispensed: 0, rejected: 0 }
        },
        'Cassette 4': { 
          supply: 'Good state', 
          fitness: 'No error',
          counters: { loaded: 0, dispensed: 0, rejected: 0 }
        },
        'Statement Paper': { supply: 'Not configured', fitness: 'No error' },
        'Statement Ribbon': { supply: 'Not configured', fitness: 'No error'},
        'Envelope Dispenser': { supply: 'Not configured', fitness: 'No error'}
      }
      expect(hw.devices).toEqual(devices);
    });
  });

  describe('getDeviceSupply()', function(){
    beforeEach(function() {
      hw.init();
    });

    it('should get Card Capture Bin deive status', function(){
      expect(hw.getDeviceSupply('Card Capture Bin')).toEqual('Not configured');
    });
  });


  describe('setDeviceSupply()', function(){
    beforeEach(function() {
      hw.init();
    });

    it('should set supply status string', function(){
      expect(hw.getDeviceSupply('Cassette 1')).toEqual('Good state');
      expect(hw.setDeviceSupply('Cassette 1', 'Overfill')).toBeTruthy();
      expect(hw.getDeviceSupply('Cassette 1')).toEqual('Overfill');
    });
  });


  describe('getSuppliesStatus()', function(){
    beforeEach(function() {
      hw.init();
    });

    it('should get supply status string', function(){
      expect(hw.getSuppliesStatus().length).toEqual(26);
      expect(hw.getSuppliesStatus()).toEqual('00000000000000011110000000');

      expect(hw.setDeviceSupply('Cassette 1', 'Media low')).toBeTruthy();
      expect(hw.setDeviceSupply('Cassette 2', 'Not configured')).toBeTruthy();
      expect(hw.setDeviceSupply('Cassette 3', 'Media out')).toBeTruthy();
      expect(hw.setDeviceSupply('Cassette 4', 'Overfill')).toBeTruthy();
      expect(hw.getSuppliesStatus()).toEqual('00000000000000020340000000');
    });
  });

  describe('setDeviceFitness()', function(){
    beforeEach(function() {
      hw.init();
    });

    it('should set device fitness severity value', function(){
      expect(hw.getDeviceFitness('Journal Printer')).toEqual('No error');
      expect(hw.setDeviceFitness('Journal Printer', 'Routine errors have occurred')).toBeTruthy();
      expect(hw.getDeviceFitness('Journal Printer')).toEqual('Routine errors have occurred');
    })
  });

  describe('getHardwareFitness()', function(){
    beforeEach(function() {
      hw.init();
    });

    it('should get hardware fitness string', function(){
      expect(hw.getHardwareFitness().length).toEqual(38);
      expect(hw.getHardwareFitness()).toEqual('00000000000000000000000000000000000000');

      expect(hw.setDeviceFitness('Magnetic Card Reader', 'Fatal error condition exists')).toBeTruthy();
      expect(hw.setDeviceFitness('Journal Printer', 'Routine errors have occurred')).toBeTruthy();
      expect(hw.setDeviceFitness('Cash Handler', 'Warning conditions have occurred')).toBeTruthy();
      expect(hw.setDeviceFitness('Cassette 1', 'Fatal error condition exists')).toBeTruthy();
      expect(hw.setDeviceFitness('Cassette 2', 'Suspend')).toBeTruthy();
      expect(hw.setDeviceFitness('Cassette 3', 'Warning conditions have occurred')).toBeTruthy();
      expect(hw.setDeviceFitness('Cassette 4', 'Routine errors have occurred')).toBeTruthy();
      expect(hw.setDeviceFitness('Bunch Note Acceptor', 'Suspend')).toBeTruthy();
      
      expect(hw.getHardwareFitness()).toEqual('00042001000000043210000000000000003000');
    })
  });

  describe('getReleaseNumber()', function(){
    it('should get release number', function(){
      expect(hw.getReleaseNumber()).toEqual('030300');
    })
  });

  describe('getHarwareID()', function(){
    it('should get NDC hardware ID', function(){
      expect(hw.getHarwareID()).toEqual('G531‐0283');
    })
  });
});


atmhardware.js
function ATMHardwareService(){
  
  this.devices = {};

  /**
   * [init description]
   * @return {[type]} [description]
   */
  this.init = function(){
    [
      'High Order Communication',
      'System Disk',
      'Magnetic Card Reader',
      'Cash Handler',
      'Depository',
      'Encryptor',
      'Security Camera',
      'Door Access',
      'Flex Disk',
      'Statement Printer',
      'Signage Display',
      'System Display',
      'Media Entry Indicators',
      'Coin Dispensing Module Tamper Indication',
      'Voice Guidance System',
      'Bunch Note Acceptor',
      'Cheque Processor',
      'Card Capture Bin',
      'Cash Handler Reject Bin',
      'Deposit Bin',
      'Receipt Printer',
      'Journal Printer',
      'Night Safe',
      'Cassette 1',
      'Cassette 2',
      'Cassette 3',
      'Cassette 4',
      'Statement Paper',
      'Statement Ribbon',
      'Envelope Dispenser'
    ].forEach(device => {
      switch(device){
        case 'Cassette 1':
        case 'Cassette 2':
        case 'Cassette 3':
        case 'Cassette 4':
          this.devices[device] = { supply: 'Good state', fitness: 'No error' };
          break;
        default:
          this.devices[device] = { supply: 'Not configured', fitness: 'No error' };
      }
    });

    this.initCassetteCounters();
  };

  /**
   * [initCassetteCounters description]
   * @param  {[type]} loaded    [description]
   * @param  {[type]} dispensed [description]
   * @param  {[type]} rejected  [description]
   * @return {[type]}           [description]
   */
  this.initCassetteCounters = function(loaded, dispensed, rejected){
    if(loaded === undefined)
      loaded = 0;

    if(dispensed === undefined)
      dispensed = 0;

    if(rejected === undefined)
      rejected = 0;

    [ 'Cassette 1', 'Cassette 2', 'Cassette 3', 'Cassette 4' ].forEach(cassette => {
      this.devices[cassette].counters = { 'loaded': loaded, 'dispensed': dispensed, 'rejected': rejected };
    });
  };

  /**
   * [setDeviceSupply description]
   * @param {[type]} device_name [description]
   * @param {[type]} supply      [description]
   */
  this.setDeviceSupply = function(device_name, supply){
    var device = this.devices[device_name];
    if(device){
      switch(supply){
        case 'Not configured':
        case 'Good state':
        case 'Media low':
        case 'Media out':
        case 'Overfill':
          device.supply = supply;
          break;
        default:
          console.log('Unsupported supply status value: ' + supply);
          break;
      }

      return true;
    }

    return false;
  };

  /**
   * [getDeviceSupply description]
   * @param  {[type]} name [description]
   * @return {[type]}      [description]
   */
  this.getDeviceSupply = function(name){
    var device = this.devices[name];

    if(device){
      return device.supply;
    }
  };

  /**
   * [setDeviceFitness description]
   * @param {[type]} device_name [description]
   * @param {[type]} fitness     [description]
   */
  this.setDeviceFitness = function(device_name, fitness){
    var device = this.devices[device_name];
    if(device){
      switch(fitness){
        case 'No error':
        case 'Routine errors have occurred':
        case 'Warning conditions have occurred':
        case 'Suspend':
        case 'Fatal error condition exists':
          device.fitness = fitness;
          break;
        default:
          console.log('Unsupported fitnes severity value: ' + fitness);
          break;
      }

      return true;
    }

    return false;
  };


  /**
   * [getDeviceFitness description]
   * @param  {[type]} name [description]
   * @return {[type]}      [description]
   */
  this.getDeviceFitness = function(name){
    var device = this.devices[name];

    if(device){
      return device.fitness;
    }
  };

  /**
   * [getSuppliesStatus description]
   * @return {[type]} [description]
   */
  this.getSuppliesStatus = function(){
    var status = '';

    [
      'Reserved',
      'Reserved',
      'Not used',
      'Card Capture Bin',
      'Cash Handler Reject Bin',
      'Deposit Bin',
      'Receipt Printer',
      'Journal Printer',
      'Not used',
      'Not used',
      'Night Safe',
      'Not used',
      'Not used',
      'Not used',
      'Not used',
      'Cassette 1',
      'Cassette 2',
      'Cassette 3',
      'Cassette 4',
      'Not used',
      'Not used',
      'Statement Paper',
      'Statement Ribbon',
      'Reserved',
      'Reserved',
      'Envelope Dispenser'
    ].forEach(device => {
      switch(device){
        case 'Reserved':
        case 'Not used':
          status += '0';
          break;

        default: 
          switch(this.getDeviceSupply(device)){
            case 'Not configured':
              status += '0';
              break;

            case 'Good state':
              status += '1';
              break;

            case 'Media low':
              status += '2';
              break;

            case 'Media out':
              status += '3';
              break;

            case 'Overfill':
              status += '4';
              break;

            default:
              status += '0';
              break;
          }
          break;
      }
    });

    return status;
  };

  this.getHardwareFitness = function(){
    var hw_fitness = '';
    [
      'Time‐of‐Day Clock',
      'High Order Communication',
      'System Disk',
      'Magnetic Card Reader',
      'Cash Handler',
      'Depository',
      'Receipt Printer',
      'Journal Printer',
      'Reserved',
      'Reserved',
      'Night Safe',
      'Encryptor',
      'Security Camera',
      'Door Access',
      'Flex Disk',
      'Cassette 1',
      'Cassette 2',
      'Cassette 3',
      'Cassette 4',
      'Reserved',
      'Reserved',
      'Statement Printer',
      'Signage Display',
      'Reserved',
      'Reserved',
      'System Display',
      'Media Entry Indicators',
      'Envelope Dispenser',
      'Not supported',
      'Coin Dispensing Module Tamper Indication',
      'Not supported',
      'Reserved',
      'Voice Guidance System',
      'Reserved',
      'Bunch Note Acceptor',
      'Cheque Processor',
      'Not supported',
      'Reserved',
    ].forEach(device => {
      var fitness = this.getDeviceFitness(device);
      switch(fitness){
        case 'No error':
          hw_fitness += '0';
          break;

        case 'Routine errors have occurred':
          hw_fitness += '1';
          break;

        case 'Warning conditions have occurred':
          hw_fitness += '2';
          break;

        case 'Suspend':
          hw_fitness += '3';
          break;

        case 'Fatal error condition exists':
          hw_fitness += '4';
          break;
          
        default:
          hw_fitness += '0';
          break;
      }
    });

    return hw_fitness;
  };

  this.getSuppliesDescription = function(){
    var description = {};
  };


  this.getReleaseNumber = function(){
    return '030300';
  }

  this.getHarwareID = function(){
    return 'G531‐0283';
  };
}

module.exports = ATMHardwareService;